src/groupeddataframe/splitapplycombine.jl

```diff
@@ -523,58 +541,66 @@ function _combine(gd::GroupedDataFrame,
     # and if a given column can be replaced in the future
     seen_cols = Dict{Symbol, Tuple{Bool, Int}}()
 
+    tasks = similar(cs_norm, Task)
+
     parentdf = parent(gd)
-    for i in eachindex(cs_norm, optional_transform)
+    for i in eachindex(cs_norm, optional_transform, tasks)
         cs_i = cs_norm[i]
         optional_i = optional_transform[i]
 
-        if length(gd) > 0 && isagg(cs_i, gd)
+        tasks[i] = Threads.@spawn if length(gd) > 0 && isagg(cs_i, gd)
             _combine_process_agg(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
         elseif keeprows && cs_i isa Pair && first(last(cs_i)) === identity &&
                !(first(cs_i) isa AsTable) && (last(last(cs_i)) isa Symbol)
             # this is a fast path used when we pass a column or rename a column in select or transform
             _combine_process_noop(cs_i, optional_i, parentdf, seen_cols, trans_res, idx_keeprows, copycols)
         elseif cs_i isa Base.Callable
-            idx_callable = _combine_process_callable(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
-            if idx_callable !== nothing
-                if idx_agg === nothing
-                    idx_agg = idx_callable
-                else
-                    @assert idx_agg === idx_callable
-                end
-            end
+            _combine_process_callable(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
         else
             @assert cs_i isa Pair
-            idx_pair = _combine_process_pair(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
-            if idx_pair !== nothing
-                if idx_agg === nothing
-                    idx_agg = idx_pair
-                else
-                    @assert idx_agg === idx_pair
-                end
+            _combine_process_pair(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
+        end
+    end
+    for t in tasks
+        local postprocessf
+        try
+            postprocessf = fetch(t)
+        catch e
+            if e isa TaskFailedException
+                rethrow(t.exception)
+            else
+                rethrow()
+            end
+        end
+        idx = postprocessf()
```

Comment:
I think given the `Ref` logic you have implemented we:
1. do not need this step with `idx` (it is enough to call `postprocessf()`)
2. we do not have to return `idx` from the anonymous functions above (you correctly update `idx_agg` in the codes with `lock` part - this is a side benefit of having `Ref` here)

Agreed?

Reply:
Yes, good catch. At least I hope so. Do you think there's anything to check to ensure that's the case (beyond seeing tests pass)?

Reply:
The tests I proposed in https://github.com/JuliaData/DataFrames.jl/pull/2574#issuecomment-739541672 are intended to make sure that all works correctly.
+
we have to double check each update and use of `idx_agg` is consistent (I did this and it looks all is fine, but a second pair of eyes would not hurt - but I assume you also already did this)

Reply:
OK I've added a test. Yes I tried to check, but I'm not super familiar with the code logic.

Reply:
I will re-check this when the CI starts to pass and let you know.